contributors/design-proposals/auth/self-hosting-authorization-webhook.md

```diff
@@ -0,0 +1,369 @@
+
+- Kubeconfig files already have a widely accepted use, and extending them to
+  make services first-class citizens is a non-goal.
+
+### Using a new configuration flag
+
+Versus: overloading the configuration flag with a new meaning.
+
+More flags mean a larger configuration space.  It seems beneficial to avoid
+introducing new flags if possible.  Reusing a flag also makes it obvious that
+there may only be one authz configuration active at a time.
+
+### Using dynamic authorizator configuration
+
+Versus: static authorizator configuration in a file.
+
+[Concerns were raised][11] about the ability of the cluster admin to
+misconfigure a cluster, or undo a previously established hosted configuration
+if dynamic configuration is used.
+
+This proposal side-steps that concern as it still relies on the use of a
+configuration file, which is out of reach of cluster admins in hosted solutions.
+At the same time, it does not prevent a future change that would, if required,
+admit dynamic configuration.
+
+### Allowing a nonempty request path with service resolution
+
+Versus: always requesting on `https://serviceIP:servicePort/`
+
+While initial version of this proposal contained a way to send a request to
+arbitrary path of the resolved service hostport, expressing that ended up being
+somewhat akward, requiring technically correct, but uncommon constructions like:
+
+```yaml
+# Much elided
+# ...
+    service:
+      name: some-service
+      namespace: some-namespace
+    url: https:/some/path
+```
+
+Suggestions that would allow setting the request path elegantly gladly accepted.
+
+### Using PEM-encoded configuration in the configuration file
+
+Versus: specifying filesystem paths to the certificates.
+
+Specifying the filesystem paths allows the authz configuration file to remain
+the same across deployments, while only the mounted contents of the needed files
+vary where it applies.
+
+Contrast that to self-registration, where certificates and keys would have to be
+passed along as inline content in the API request to register a webhook.
+
+Since we are currently not considering dynamic registration (see previous
+subsection), we can accept only having fields that point at filesystem paths.
+This decision does not prevent changing the direction in a followup version, say
+by allowing filesystem references and inline content to coexist.
+
+### Extending generic apiserver
+
+Versus: extending kube-apiserver only.
+
+Only the `kube-apiserver` has the ability to resolve service endpoints without
+special configuration.  Also, `kube-apiserver` is the only apiserver that has
+the use case described in this document.
+
+Therefore, it seems reasonable to confine the behavior changes to
+`kube-apiserver` alone.
+
+### Unifying code paths for webhooks
+
+Versus: this work is out of scope.
+
+Our proposed solution makes no change to the current state of the webhook
+libraries.  Unifying code paths for webhooks is a worthwhile goal, but out of
```

Comment:
If we're going to be adding features to the authorization webhook, it seems odd to declare "how other webhooks will us this feature it's out of scope." I'd like to see it at least considered, even if we don't implemented.

Reply:
Other webhooks won't use this feature per this proposal.  The proposal (at least IMHO) doesn't prevent reuse.

I can expand a bit on that viewpoint in the text, but anything I'd say would be along the lines of "this is not a concern in the context of this proposal".

Reply:
Also, forgot to mention: Thanks a million for jumping to review this pull request so quickly!

Reply:
I seflishly would like to see this usable by the authentication webhook too. Looking at https://github.com/kubernetes/kubernetes/pull/54733 it doesn't seem like that would be too difficult, the language here just worries me.

Reply:
Please take another look.  I tried to clarify.